sql/core/src/test/scala/org/apache/spark/sql/streaming/StreamingQueryHashPartitionVerifySuite.scala

```diff
@@ -0,0 +1,199 @@
+    (rows, partIds)
+  }
+
+  private def getRandomRows(schema: StructType, numRows: Int, rand: Random): Seq[UnsafeRow] = {
+    val generator = RandomDataGenerator
+      .forType(schema, rand = new Random(rand.nextInt(RandomDataGenerator.MAX_STR_LEN)))
+      .get
+
+    // Create the converters needed to convert from external row to internal
+    // row and to UnsafeRows. Projection itself costs a lot on initialization
+    // (codegen and compile), so initialize it once.
+    val internalConverter = CatalystTypeConverters.createToCatalystConverter(schema)
+    val unsafeConverter = UnsafeProjection.create(Array(schema).asInstanceOf[Array[DataType]])
+
+    (1 to numRows).map { _ =>
+      val row = generator().asInstanceOf[Row]
+
+      val internalRow = new GenericInternalRow(1)
+      internalRow.update(0, internalConverter(row).asInstanceOf[GenericInternalRow])
+      val unsafeRow = unsafeConverter.apply(internalRow)
+      unsafeRow.copy()
+    }
+  }
+
+  private def getRandomSchemas(numSchemaTypes: Int, rand: Random,
+                               maxNumFields: Int): Seq[StructType] = {
+    (1 to numSchemaTypes).map { _ =>
+      val numField = rand.nextInt(maxNumFields)
+      RandomDataGenerator.randomSchema(rand, Math.max(1, numField), variableTypes)
+    }
+  }
+
+  private def getPartitionId(rows: Seq[UnsafeRow], hash: HashPartitioning): Seq[Int] = {
+    val partIdExpr = hash.partitionIdExpression
+    rows.map{row =>
+      partIdExpr.eval(row).asInstanceOf[Int]
+    }
+  }
+
+  test("SPARK-47788: Ensure the same hash function is used across batches.") {
+    val rowAndPartIdFile = new File(goldenFileURI.getPath, rowAndPartIdFilename)
+
+    if (regenerateGoldenFiles) {
+      val seed = Random.nextInt()
+      val random = new Random(seed)
+      logInfo(s"Get random inputs with seed $seed")
+
+      val schemas = getRandomSchemas(numSchemaTypes, random, maxNumFields)
+
+      val os = new DataOutputStream(
+        codec.compressedOutputStream(new FileOutputStream(rowAndPartIdFile))
+      )
+
+      saveSchemas(schemas)
+      schemas.foreach { schema =>
+          // Streaming stateful ops rely on this distribution to partition the data.
+          // Spark should make sure this class's partition dependency remain unchanged.
+          val hash = StatefulOpClusteredDistribution(
+            Seq(BoundReference(0, schema, nullable = true)),
+            numShufflePartitions
+          ).createPartitioning(numShufflePartitions)
+
+          assert(
+            hash.isInstanceOf[HashPartitioning],
+            "StatefulOpClusteredDistribution should " +
+            "rely on HashPartitioning to ensure partitions remain the same for streaming " +
+            "stateful operators."
+          )
+          val rows = getRandomRows(schema, numRows, random)
+          val partitions = getPartitionId(rows, hash.asInstanceOf[HashPartitioning])
+          saveRowsAndPartIds(rows, partitions, os)
```

Comment:
So in the binary golden file, the format is like (no new line, added for better understanding):
```
<num rows>
row1 part1
row2 part2
...
<num rows>
row1 part1
row2 part2
...
```

If so, could you add a comment here explaining this layout? Thanks in advance!

Reply:
The layout is not for human readable since the rows will be compressed (and without space and newline), so instead of visualizing the layout, added a few comments.